services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -0,0 +1,326 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.protobuf.service;
+
+import com.google.protobuf.Descriptors.Descriptor;
+import com.google.protobuf.Descriptors.FieldDescriptor;
+import com.google.protobuf.Descriptors.FileDescriptor;
+import com.google.protobuf.Descriptors.MethodDescriptor;
+import com.google.protobuf.Descriptors.ServiceDescriptor;
+
+import io.grpc.NotifyOnServerBuild;
+import io.grpc.Server;
+import io.grpc.ServerServiceDefinition;
+import io.grpc.Status;
+import io.grpc.protobuf.ProtoFileDescriptorSupplier;
+import io.grpc.reflection.v1alpha.ErrorResponse;
+import io.grpc.reflection.v1alpha.ExtensionNumberResponse;
+import io.grpc.reflection.v1alpha.ExtensionRequest;
+import io.grpc.reflection.v1alpha.FileDescriptorResponse;
+import io.grpc.reflection.v1alpha.ListServiceResponse;
+import io.grpc.reflection.v1alpha.ServerReflectionGrpc;
+import io.grpc.reflection.v1alpha.ServerReflectionRequest;
+import io.grpc.reflection.v1alpha.ServerReflectionResponse;
+import io.grpc.reflection.v1alpha.ServiceResponse;
+import io.grpc.stub.StreamObserver;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+
+/**
+ * Provides a reflection service for Protobuf services (including the reflection service itself).
+ */
+public final class ProtoReflectionService extends ServerReflectionGrpc.ServerReflectionImplBase
+    implements NotifyOnServerBuild {
+  private Server server;
+  private Set<String> serviceNames;
+  private Map<String, FileDescriptor> fileDescriptorsByName;
+  private Map<String, FileDescriptor> fileDescriptorsBySymbol;
+  private Map<String, Map<Integer, FileDescriptor>> fileDescriptorsByExtensionAndNumber;
+  private Boolean mapsInitialized = false;
+
+  /**
+   * Retrieves the services registered to the server at build time, and stores those services with a
+   * {@link ProtoFileDescriptorSupplier}.
+   */
+  public void notifyOnBuild(Server server) {
+    this.server = server;
+  }
+
+  private void processExtension(FieldDescriptor extension, FileDescriptor fd) {
+    String extensionName = extension.getContainingType().getFullName();
+    int extensionNumber = extension.getNumber();
+    if (fileDescriptorsByExtensionAndNumber.containsKey(extensionName)) {
+      fileDescriptorsByExtensionAndNumber.get(extensionName).put(extensionNumber, fd);
+    } else {
+      Map<Integer, FileDescriptor> extensionMap = new HashMap<Integer, FileDescriptor>();
+      extensionMap.put(extensionNumber, fd);
+      fileDescriptorsByExtensionAndNumber.put(extensionName, extensionMap);
+    }
+  }
+
+  private void processService(ServiceDescriptor service, FileDescriptor fd) {
+    fileDescriptorsBySymbol.put(service.getFullName(), fd);
+    for (MethodDescriptor method : service.getMethods()) {
+      fileDescriptorsBySymbol.put(method.getFullName(), fd);
```

Comment:
I would not make that assumption.  There is very little enforcement of that, if at all.  David's comments are borne from C++, which treats all protos as part of the binary and thus may do such deduplication.  Java does not.